kernel_tracing/metric_evaluation/test.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright(c) 2022 Intel Corporation. All rights reserved.
+
+# Run with --help for a description of this script
+
+import os
+import subprocess
+import re
+import argparse
+import pytest
+import json
+
+
+class BPF:
+    def __init__(self, bpftrace_script):
+        self.script = bpftrace_script
+        self.process = None
+
+    def start(self):
+        self.process = subprocess.Popen(
+            ["bpftrace", self.script], stdout=subprocess.PIPE, encoding="utf-8")
+        # Wait for bpftrace script to load
+        while True:
+            line = self.process.stdout.readline()
+            if "Attaching" in line:
+                break
+
+    def stop(self):
+        # Send ctrl-c
+        self.process.send_signal(subprocess.signal.SIGINT)
+        self.process.wait()
+
+        # Collect printed variables for condition evaluation
+        lines = self.process.stdout.read().split("\n")
+        bpftrace_vars = {}
+        for line in lines:
+            # Check if line is outputting a variable from the script
+            # bpftrace output format is `@name[key1, key2, ...] = val` (keys are optional)
+            if re.match("@.*: .*", line):
+                key, val = line.split(": ", 1)
+                bpftrace_vars[key] = val
+        return bpftrace_vars
+
+
+def check_conditions(conditions, bpftrace_vars):
+    failedConditions = []
+    for condition in conditions:
+        conditionWithVars = condition
+        for key, val in bpftrace_vars.items():
+            conditionWithVars = conditionWithVars.replace(key, val)
+        # Eval works fine here as the spec file already has RCE via the shell parameter
+        # pylint: disable=eval-used
+        try:
+            if not eval(conditionWithVars):
+                failedConditions.append(
+                    f"Failed condition: {condition}. Evaluated to {conditionWithVars}.")
+        except SyntaxError as err:
+            failedConditions.append(
+                f"Invalid condition: {err.text}. Are you sure the bpftrace script outputs necessary variables?")
+
+    if len(failedConditions) != 0:
+        pytest.fail(
+            '\n'.join([*failedConditions, f"bpftrace vars:\n{json.dumps(bpftrace_vars, indent=2)} "]), pytrace=False)
+
+
+def test(test_case, working_dir):
+    """ This gets called by pytest, args are generated in conftest.py """
+    bpftrace_script = os.path.join(working_dir, test_case['script'])
+    shell_commands = test_case['shell']
+    conditions = test_case['conditions']
+    if isinstance(shell_commands, str):
+        shell_commands = [shell_commands]
+
+    # Start bpftrace collection
+    bpf = BPF(bpftrace_script)
+    bpf.start()
```

Comment:
Starting and stopping the BPF trace is not part of the test but part of the test "fixture" https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#fixtures-can-be-requested-more-than-once-per-test-return-values-are-cached

Reply:
More relevant link sorry:
https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#scope-sharing-fixtures-across-classes-modules-packages-or-session